org.eclipse.m2e.pde.connector/src/org/eclipse/m2e/pde/connector/Activator.java

```diff
@@ -0,0 +1,44 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Christoph Läubrich
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *      Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.m2e.pde.connector;
+
+import java.util.Optional;
+
+import org.eclipse.pde.core.project.IBundleProjectService;
+import org.osgi.framework.BundleActivator;
+import org.osgi.framework.BundleContext;
+import org.osgi.util.tracker.ServiceTracker;
+
+public class Activator implements BundleActivator {
```

Comment:
I suggest to use the FrameworkUtil method to access OSGI services instead of an activator. Easier to read and avoids code.

Reply:
I don't see anything `org.osgi.framework.FrameworkUtil` can do here... If you mean `org.eclipse.core.runtime.ServiceCaller<Service>` this class introduces bad OSGi practice and I won't suggest using it anywhere at all...

Reply:
FrameworkUtil.getBundle(this.getClass()).getBundleContext() can be used to access the bundle context and afterwards the service. via it.

Reply:
In platform code we typically do not hold on to OSGi services but use util classes to access it dynanically to the listener is not needed. But having this activator is also not a really bad thing, if you prefer this pattern that is also fine of course. Activators is something which tend to grow over time and if large can slow down the startup of the application hence we try to avoid using them in platform code.